### PR TITLE
[Refactor] Extract common logics to function success_once

### PR DESCRIPTION
**Pull request by sduzh**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [ ] enhancement
- [x] refactor
- [ ] others

### CODE REVIEW

File: be/src/util/once.h [hunk 1]
```diff
@@ -1,78 +1,54 @@
...
-//     bool is_inited() const {
-//       return _init_once.has_called() && _init_once.stored_result().ok();
-//     }
-//   private:
-//     Status _do_init() { /* init logic here */ }
-//     StarRocksCallOnce<Status> _init_once;
-//   };
-template <typename ReturnType>
-class StarRocksCallOnce {
+class SuccessOnce {
```
Comment (imay) [inline be/src/util/once.h:14]: Could you add some comments for this class to explain why and how to use it.

File: be/src/util/once.h [hunk 2]
```diff
@@ -1,78 +1,54 @@
...
-        // std::memory_order_acquire here and std::memory_order_release in
-        // init(), taken together, mean that threads can safely synchronize on
-        // _has_called.
-        return _has_called.load(std::memory_order_acquire);
-    }
+    // Return value:
+    //  true: |fn| is invoked and returned an OK status.
+    //  false: |fn| has been invoked before and returned an OK status.
+    //  non-OK status: |fn| is invoked and returned a non-OK status.
+    StatusOr<bool> call(const std::function<Status()>& fn);
```
Comment (stdpain) [inline be/src/util/once.h:22]: how about using a template function?
```
template<class fn>
StatusOr<bool> call(fn&& func) {

```

File: be/src/util/once.h [hunk 3]
```diff
@@ -1,78 +1,54 @@
...
 private:
-    std::atomic<bool> _has_called;
-    std::once_flag _once_flag;
-    ReturnType _status;
+    std::atomic<int> _flag;
 };
 
+inline StatusOr<bool> SuccessOnce::call(const std::function<Status()>& fn) {
+    while (true) {
+        auto curr_flag = _flag.load(std::memory_order_acquire);
```
Comment (stdpain) [inline be/src/util/once.h:32]: Why not use std::call_once directly

File: be/src/util/once.h (hunk 1 above)
Comment (sduzh) [reply be/src/util/once.h]: updated

File: be/src/util/once.h (hunk 2 above)
Comment (sduzh) [reply be/src/util/once.h]: updated

File: be/src/util/once.h (hunk 3 above)
Comment (sduzh) [reply be/src/util/once.h]: Updated, please check the comments above the function `success_once`.

**Review by stdpain — APPROVED** [review]

**Review by imay — APPROVED** [review]

**Comment by sduzh** [comment]
@mergify backport branch-2.3

**Pull request merged**